neo/Consensus/RecoveryMessage.cs

```diff
@@ -1,20 +1,20 @@
-﻿using System.IO;
-using Neo.IO;
+﻿using Neo.IO;
+using Neo.Ledger;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
 
 namespace Neo.Consensus
 {
-    internal class RecoveryMessage : ConsensusMessage
+    internal partial class RecoveryMessage : ConsensusMessage
     {
-        public byte[][] ChangeViewWitnessInvocationScripts;
-        public uint[] ChangeViewTimestamps;
-        public byte[] OriginalChangeViewNumbers;
+        public Dictionary<ushort, ChangeViewPayloadCompact> ChangeViewMessages;
         public PrepareRequest PrepareRequestMessage;
-
         /// The PreparationHash in case the PrepareRequest hasn't been received yet.
         /// This can be null if the PrepareRequest information is present, since it can be derived in that case.
         public UInt256 PreparationHash;
-        public byte[][] PrepareWitnessInvocationScripts;
-        public byte[][] CommitSignatures;
+        public Dictionary<ushort, PreparationPayloadCompact> PreparationMessages;
+        public Dictionary<ushort, CommitPayloadCompact> CommitMessages;
```

Comment:
Nitpick:
If you leave these as an array you don't have to have the `ValidatorIndex` in the objects themselves. Also, switching to serialize with the index doesn't really save that much space, because It was already serializing them sparsely before this change by just including a single `0x00` byte for entries not present. With less entries present it may save a little space here, but with more it uses a little more space.

Even if you want really want to keep the `ValidatorIndex` in the objects and serialize them sparsely with the index, you could still use an array here to hold them instead of a Dictionary and it is more optimal.

I'm sure when profiling it would be very little of where the time is spent anyway, though. I'll approve the PR and leave it up to you to decide.